File: src/lib/api/google.ts
// ─────────────────────────────────────────────────────────────────────────────
// Olive Mode — Google Trends API Service
// Place at: src/lib/api/google.ts
// ─────────────────────────────────────────────────────────────────────────────

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

// ─── Types ────────────────────────────────────────────────────────────────────

export interface GoogleMomentum {
  current_avg:  number;
  previous_avg: number;
  change_pct:   number;
  direction:    "up" | "down" | "flat";
}

export interface GoogleDataPoint {
  date:   string;
  values: Record<string, number>;
}

export interface GoogleKeywordResponse {
  keywords:  string[];
  timeframe: string;
  geo:       string;
  found:     boolean;
  momentum:  Record<string, GoogleMomentum>;
  data:      GoogleDataPoint[];
}

export interface GoogleRelatedQuery {
  query: string;
  value: number;
}

export interface GoogleRelatedResponse {
  keyword: string;
  geo:     string;
  rising:  GoogleRelatedQuery[];   // fast-growing — early trend signals
  top:     GoogleRelatedQuery[];   // highest volume
}

export interface GoogleRegion {
  location: string;
  interest: number;
}

export interface GoogleRegionalResponse {
  keyword:    string;
  geo:        string;
  resolution: string;
  regions:    GoogleRegion[];
}

export interface GoogleCategoryItem {
  category:     string;
  avg_interest: number;
  peak:         number;
  trend:        number[];
}

export interface GoogleCompareResponse {
  geo:        string;
  timeframe:  string;
  categories: GoogleCategoryItem[];
}

// ─── Keyword trend over time ──────────────────────────────────────────────────

export async function fetchGoogleKeywordTrend(
  keywords:  string | string[],
  timeframe  = "today 3-m",
  geo        = "US"
): Promise<GoogleKeywordResponse | null> {
  try {
    const kw = Array.isArray(keywords) ? keywords.join(",") : keywords;
    const params = new URLSearchParams({ keywords: kw, timeframe, geo });
    const response = await fetch(
      `${API_BASE}/google/trends/keyword?${params.toString()}`
    );
    if (!response.ok) return null;
    const data = await response.json() as GoogleKeywordResponse;
    console.log("[google] raw response:", JSON.stringify(data, null, 2));
    return data;
  } catch {
    console.error("[google] fetchGoogleKeywordTrend failed");
    return null;
  }
}

// ─── Related queries ──────────────────────────────────────────────────────────

export async function fetchGoogleRelatedQueries(
  keyword: string,
  geo      = "US"
): Promise<GoogleRelatedResponse | null> {
  try {
    const params = new URLSearchParams({ keyword, geo });
    const response = await fetch(
      `${API_BASE}/google/trends/related?${params.toString()}`
    );
    if (!response.ok) return null;
    return response.json() as Promise<GoogleRelatedResponse>;
  } catch {
    console.error("[google] fetchGoogleRelatedQueries failed");
    return null;
  }
}

// ─── Regional interest ────────────────────────────────────────────────────────

export async function fetchGoogleRegionalInterest(
  keyword:    string,
  geo         = "US",
  resolution  = "REGION"
): Promise<GoogleRegionalResponse | null> {
  try {
    const params = new URLSearchParams({ keyword, geo, resolution });
    const response = await fetch(
      `${API_BASE}/google/trends/regional?${params.toString()}`
    );
    if (!response.ok) return null;
    return response.json() as Promise<GoogleRegionalResponse>;
  } catch {
    console.error("[google] fetchGoogleRegionalInterest failed");
    return null;
  }
}

// ─── Category comparison ──────────────────────────────────────────────────────

export async function fetchGoogleCategoryCompare(
  geo       = "US",
  timeframe = "today 3-m"
): Promise<GoogleCompareResponse | null> {
  try {
    const params = new URLSearchParams({ geo, timeframe });
    const response = await fetch(
      `${API_BASE}/google/trends/compare?${params.toString()}`
    );
    if (!response.ok) return null;
    return response.json() as Promise<GoogleCompareResponse>;
  } catch {
    console.error("[google] fetchGoogleCategoryCompare failed");
    return null;
  }
}

File: agent_backend/routers/google.py
# ─────────────────────────────────────────────────────────────────────────────
# Olive Mode — Google Trends Router
# File: routers/google.py
# ─────────────────────────────────────────────────────────────────────────────

from fastapi import APIRouter, HTTPException, Query
from services.pytrends_client import (
    get_pytrends,
    build_payload,
    interest_over_time,
    calculate_momentum,
    shape_related,
    clean_keyword,
)

router = APIRouter(prefix="/google", tags=["Google Trends"])

# Olive Mode's core product categories for comparison
OLIVE_MODE_CATEGORIES = [
    "mini dress",
    "midi dress",
    "crossbody bag",
    "summer jewelry",
    "hair perfume",
]


@router.get("/trends/keyword")
async def google_keyword_trend(
    keywords:  str = Query(description="Comma-separated keywords, max 5"),
    timeframe: str = Query(default="today 3-m"),
    geo:       str = Query(default="US"),
    category:  int = Query(default=185, description="185 = Fashion & Style"),
):
    """Interest over time for up to 5 keywords with momentum calculation."""
    try:
        raw_list = [k.strip() for k in keywords.split(",")][:5]
        kw_list  = [clean_keyword(k) for k in raw_list]
        print(f"[google] keywords cleaned: {raw_list} → {kw_list}")
        pt       = get_pytrends()
        build_payload(pt, kw_list, timeframe, geo, category)
        df       = interest_over_time(pt)

        if df.empty:
            return {"keywords": kw_list, "timeframe": timeframe, "geo": geo, "found": False, "data": []}

        # Only return data points with non-zero values
        data = [
            {"date": str(date)[:10], "values": {kw: int(row[kw]) for kw in kw_list if kw in row}}
            for date, row in df.iterrows()
            if any(int(row[kw]) > 0 for kw in kw_list if kw in row)
        ]

        return {
            "keywords":  kw_list,
            "timeframe": timeframe,
            "geo":       geo,
            "found":     True,
            "momentum":  calculate_momentum(df, kw_list),
            "data":      data,
        }
    except HTTPException:
        raise
    except Exception as e:
        import traceback
        print(f"[google] keyword endpoint error: {str(e)}")
        print(traceback.format_exc())
        raise HTTPException(status_code=500, detail=f"Google Trends error: {str(e)}")


@router.get("/trends/related")
async def google_related_queries(
    keyword: str = Query(description="Single keyword"),
    geo:     str = Query(default="US"),
):
    """Rising and top related queries — rising list shows early trend signals."""
    try:
        clean = clean_keyword(keyword)
        print(f"[google] related cleaned: '{keyword}' → '{clean}'")
        pt      = get_pytrends()
        build_payload(pt, [clean], "today 3-m", geo)
        related = pt.related_queries()
        result  = related.get(clean, {})

        return {
            "keyword": clean,
            "geo":     geo,
            "rising":  shape_related(result.get("rising")),
            "top":     shape_related(result.get("top")),
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Google Trends error: {str(e)}")


@router.get("/trends/regional")
async def google_regional_interest(
    keyword:    str = Query(description="Keyword to check"),
    geo:        str = Query(default="US"),
    resolution: str = Query(default="REGION", description="REGION | DMA | CITY"),
):
    """Geographic breakdown — which states or cities are searching most."""
    try:
        clean = clean_keyword(keyword)
        print(f"[google] regional cleaned: '{keyword}' → '{clean}'")
        pt = get_pytrends()
        build_payload(pt, [clean], "today 3-m", geo)
        df = pt.interest_by_region(resolution=resolution, inc_low_vol=False, inc_geo_code=False)

        if df.empty:
            return {"keyword": clean, "geo": geo, "regions": []}

        df = df.sort_values(by=clean, ascending=False).head(15)

        return {
            "keyword":    clean,
            "geo":        geo,
            "resolution": resolution,
            "regions": [
                {"location": loc, "interest": int(row[clean])}
                for loc, row in df.iterrows()
                if int(row[clean]) > 0
            ],
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Google Trends error: {str(e)}")


@router.get("/trends/compare")
async def google_compare_categories(
    geo:       str = Query(default="US"),
    timeframe: str = Query(default="today 3-m"),
):
    """Compare Olive Mode's core product categories against each other."""
    try:
        pt = get_pytrends()
        build_payload(pt, OLIVE_MODE_CATEGORIES, timeframe, geo)
        df = interest_over_time(pt)

        if df.empty:
            return {"categories": [], "geo": geo, "timeframe": timeframe}

        summary = []
        for cat in OLIVE_MODE_CATEGORIES:
            if cat in df.columns:
                values = df[cat].tolist()
                summary.append({
                    "category":     cat,
                    "avg_interest": round(sum(values[-4:]) / 4, 1) if values else 0,
                    "peak":         max(values) if values else 0,
                    "trend":        values[-8:],
                })

        summary.sort(key=lambda x: x["avg_interest"], reverse=True)

        return {"geo": geo, "timeframe": timeframe, "categories": summary}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Google Trends error: {str(e)}")